### PR TITLE
Adding history

### PR DESCRIPTION
**Pull request by nana-qi**

### CODE REVIEW

File: graphql-database-manager-dynamo/src/main/java/com/fleetpin/graphql/database/manager/dynamo/DynamoDb.java [hunk 1]
```diff
@@ -258,6 +267,220 @@ public int maxBatchSize() {
...
+        var toReturn = new ArrayList<T>();
+        return client.queryPaginator(builder.build())
+        .subscribe(response -> {
+            response.items().forEach(item -> toReturn.add(new DynamoItem(historyTable, item).convertTo(mapper, queryHistory.getType())));
+        }).thenApply(__ -> {
+            return toReturn;
+        });
+    }
+    
+	private AttributeValue toId(String id) {
```
Comment (ashley-taylor) [inline graphql-database-manager-dynamo/src/main/java/com/fleetpin/graphql/database/manager/dynamo/DynamoDb.java:427]: would move toId toRevisionId toUpdatedAtId to a util class I think.

File: graphql-database-manager-dynamo/src/main/java/com/fleetpin/graphql/database/manager/dynamo/DynamoDb.java [hunk 2]
```diff
@@ -258,6 +267,220 @@ public int maxBatchSize() {
             return flattener.results(mapper, key.getQuery().getType(), Optional.ofNullable(key.getQuery().getLimit()));
         });
     }
+    
+    @Override
+    public <T extends Table> CompletableFuture<List<T>> queryHistory(DatabaseQueryHistoryKey<T> key) {
```
Comment (ashley-taylor) [inline graphql-database-manager-dynamo/src/main/java/com/fleetpin/graphql/database/manager/dynamo/DynamoDb.java:272]: would try split this into some different methods. Maybe one for each of the outer if statements. Is so long.

File: graphql-database-manager-dynamo/src/main/java/com/fleetpin/graphql/database/manager/dynamo/DynamoDbManager.java [hunk 3]
```diff
@@ -110,7 +116,7 @@ public DynamoDbManager build() {
 				idGenerator = () -> UUID.randomUUID().toString();
 			}
 
-			database = Objects.requireNonNullElse(database, new DynamoDb(mapper, tables, client, idGenerator));
+			database = Objects.requireNonNullElse(database, new DynamoDb(mapper, tables, historyTable, client, idGenerator));
```
Comment (ashley-taylor) [inline graphql-database-manager-dynamo/src/main/java/com/fleetpin/graphql/database/manager/dynamo/DynamoDbManager.java:119]: historyTable is optional.
Would not have this here.
If someone tried to perform a query though and no history table should throw a nicer error than nullpointer

File: graphql-database-manager-test/src/main/java/com/fleetpin/graphql/database/manager/test/TestDatabaseProvider.java [hunk 4]
```diff
@@ -45,23 +47,29 @@
...
         final var client = startDynamoClient(port);
+        final var streamClient = startDynamoStreamClient(port);
 
         System.setProperty("sqlite4java.library.path", "native-libs");
 
         finished = new CompletableFuture<>();
 
         final var testMethod = extensionContext.getRequiredTestMethod();
         final var organisationId = testMethod.getAnnotation(TestDatabase.class).organisationId();
+        final var withHistory = testMethod.getAnnotation(TestDatabase.class).withHistory();
```
Comment (ashley-taylor) [inline graphql-database-manager-test/src/main/java/com/fleetpin/graphql/database/manager/test/TestDatabaseProvider.java:58]: we could get rid of this withHistory setting on the annotation and just look to see if one of the parameters is .isAssignableFrom(HistoryProcessor.class)

File: graphql-database-manager-dynamo/src/main/java/com/fleetpin/graphql/database/manager/dynamo/HistoryUtil.java [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Stream;
+
+import software.amazon.awssdk.services.dynamodb.model.AttributeValue;
+import software.amazon.awssdk.services.dynamodb.model.Record;
+
+public class HistoryUtil {
+
+	public static Stream<HashMap<String, AttributeValue>> toHistoryValue(List<Record> records) {
```
Comment (ashley-taylor) [inline graphql-database-manager-dynamo/src/main/java/com/fleetpin/graphql/database/manager/dynamo/HistoryUtil.java:14]: Once thing we have missed is we want to be able to control what items have history.

I think in order to get history you should have to add an annotation 
`@History`
on the classes that extend table.
Then when we do a db put we see if the type has this annotation and if we do we add a flag to the object
history = true
that this then only will process the rows that we want to capture the history on.
We should also do the check when we go to query the type make sure its a type that has the `@History` annotation and error the query if we don't have it

File: graphql-database-manager-test/src/main/java/com/fleetpin/graphql/database/manager/test/TestDatabaseProvider.java [hunk 6]
```diff
@@ -45,23 +47,29 @@
...
-                            return createDynamoDbManager(client, parameter);
-                        } else {
-                            return createDatabase(client, parameter, organisationId);
+                            return createDynamoDbManager(client, streamClient, parameter, withHistory);
+                        } else if (parameter.getType().isAssignableFrom(HistoryProcessor.class)) {
+                        	return new HistoryProcessor(client, streamClient, parameter, organisationId);
+                        }
+                        else {
+                            return createDatabase(client, streamClient, parameter, organisationId, withHistory);
                         }
```
Comment (Aliics) [inline graphql-database-manager-test/src/main/java/com/fleetpin/graphql/database/manager/test/TestDatabaseProvider.java:70]: This is starting to get long.
Factory might be nice, not required.

File: graphql-database-manager-core/src/main/java/com/fleetpin/graphql/database/manager/QueryHistoryBuilder.java [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+
+    public QueryHistory<V> build() {
+    	Preconditions.checkArgument(!(id != null && startsWith != null), "ID and StartsWith cannot both be set.");
+    	Preconditions.checkArgument(!(id == null && startsWith == null), "ID or StartsWith must be set.");
+    	if (fromRevision != null || toRevision != null) {
+    		Preconditions.checkArgument(fromUpdatedAt == null && toUpdatedAt == null, "Revision and CreatedAt cannot both be set.");
+    		Preconditions.checkArgument(startsWith == null, "StartsWith can only be used with updatedAt.");
+    	}
+    	return new QueryHistory<V>(type, startsWith, id, fromRevision, toRevision, fromUpdatedAt, toUpdatedAt);
+    }
```
Comment (Aliics) [inline graphql-database-manager-core/src/main/java/com/fleetpin/graphql/database/manager/QueryHistoryBuilder.java:65]: I feel like these builder conditions solidify my statement on how this should actually have multiple implementations stated on a comment I have in DynamoDb.

File: graphql-database-manager-dynamo/src/main/java/com/fleetpin/graphql/database/manager/dynamo/DynamoDb.java (hunk 1 above)
Comment (Aliics) [reply graphql-database-manager-dynamo/src/main/java/com/fleetpin/graphql/database/manager/dynamo/DynamoDb.java:384]: I would have agree here :)

File: graphql-database-manager-test/src/main/java/com/fleetpin/graphql/database/manager/test/HistoryProcessor.java [hunk 8]
```diff
@@ -61,8 +61,13 @@ public void process() {
 		        	
 		        	var items = new HashMap<String, List<WriteRequest>>();
 		        	var writeRequests = stream.map(item -> WriteRequest.builder().putRequest(builder -> builder.item(item)).build()).collect(Collectors.toList());
-		        	items.put(table+"_history", writeRequests);
-		        	client.batchWriteItem(builder -> builder.requestItems(items)).get();
+		        	if (writeRequests == null || writeRequests.isEmpty()) {
```
Comment (ashley-taylor) [inline graphql-database-manager-test/src/main/java/com/fleetpin/graphql/database/manager/test/HistoryProcessor.java:64]: would not have an empty if with an else would just change the if to 

`writeRequests != null && !writeRequests.isEmpty()`

File: graphql-database-manager-test/src/test/java/com/fleetpin/graphql/database/manager/test/DynamoDbHistoryTest.java [hunk 9]
```diff
@@ -433,6 +434,23 @@ void testStartWithToUpdatedAtQuery(final Database db, final HistoryProcessor his
...
+        table1 = new NoHistorySameNameTable("revision2");
+        table1.setId("testTable1");
+        table1.setRevision(1);
+        db.put(table1).get();
+        
+        historyProcessor.process();
+		var history = db.queryHistory(QueryHistoryBuilder.create(SameNameTable.class).id("testTable1").build()).get();
+		Assertions.assertEquals(0, history.size());
+
+	}
```
Comment (ashley-taylor) [inline graphql-database-manager-test/src/test/java/com/fleetpin/graphql/database/manager/test/DynamoDbHistoryTest.java:452]: looks good would just add some tests testing the failure cases and confirming they through errors.
I.e. setting both startWith and id
trying to do a history query on an object that does not have the `@History` annotation
then good to go 👍

**Pull request merged**